src/main/java/com/minecolonies/coremod/entity/WitchPotionEntity.java

```diff
@@ -0,0 +1,134 @@
+     */
+    public static final double MAX_DISTANCE = 16.0D;
+
+    /**
+     * The minimum duration to get affected
+     */
+    public static final int MIN_DURATION = 20;
+
+    /**
+     * Throws a potion at the target with the given inaccuracy
+     *
+     * @param potionStack the {@link ItemStack} of the Potion, {@link ItemStack#getItem()} must return a {@link net.minecraft.item.ThrowablePotionItem}
+     * @param target the targeted {@link LivingEntity} to throw the potion at
+     * @param thrower the witch throwing the potion
+     * @param world the {@link World} of the thrower
+     * @param velocity the velocity to throw the potion with
+     * @param inaccuracy the inaccuracy to throw the potion with
+     */
+    public static void throwPotionAt(final ItemStack potionStack, final LivingEntity target, final AbstractEntityCitizen thrower, final World world, final float velocity, final float inaccuracy)
+    {
+        final PotionEntity potionentity = new WitchPotionEntity(world, thrower);
+        potionentity.setItem(potionStack);
+        potionentity.shoot(target.getPosX(), target.getPosY(), target.getPosZ(), velocity, inaccuracy);
+        world.addEntity(potionentity);
+    }
+
+    public WitchPotionEntity(final EntityType<? extends PotionEntity> typeIn, final World worldIn)
+    {
+        super(typeIn, worldIn);
+    }
+
+    public WitchPotionEntity(final World worldIn, final AbstractEntityCitizen livingEntityIn)
+    {
+        super(worldIn, livingEntityIn);
+    }
+
+    @Override
+    protected void func_213888_a(final List<EffectInstance> effects, @Nullable final Entity entity)
+    {
+        final AbstractEntityCitizen citizen = this.getThrower();
+        if (citizen != null && citizen.getCitizenData().getJob() instanceof JobWitch)
+        {
+            final EntityAIWitch ai = ((EntityAIWitch) citizen.getCitizenData().getJob(JobWitch.class).getWorkerAI());
+            final AxisAlignedBB axisalignedbb = this.getBoundingBox().grow(SPLASH_SIZE, SPLASH_HEIGTH, SPLASH_SIZE);
+            final List<LivingEntity> list = this.world.getEntitiesWithinAABB(LivingEntity.class, axisalignedbb);
+            if (!list.isEmpty())
+            {
+                for (final LivingEntity livingentity : list)
+                {
+                    if (livingentity.canBeHitWithPotion())
+                    {
+                        final double distanceSq = this.getDistanceSq(livingentity);
+                        if (distanceSq < MAX_DISTANCE)
+                        {
+                            double d1 = 1.0D - Math.sqrt(distanceSq) / 4.0D;
+                            if (livingentity == entity)
+                            {
+                                d1 = 1.0D;
+                            }
+                            final boolean ally = EntityAIWitch.isAlly(citizen.getCitizenData().getColony(), livingentity);
+                            final boolean foe = !ally && ai.isEntityValidTarget(livingentity);
```

Comment:
I think for the isEntityValidTarget, you should be able to calculate that from the building or so. Maybe we can even make this a static method that gets the building.

Reply:
Where should I put that static method?

Reply:
AbstractBuildingGuards I'd put it